src/protocols/bdx/BDXMessageUtils.cpp

```diff
@@ -341,62 +306,48 @@ CHIP_ERROR ReceiveAccept::Pack(System::PacketBuffer * aBuffer)
         if (mWideRange)
         {
             bbuf.PutLE64(mLength);
-            VerifyOrExit(bbuf.Fit(), err = CHIP_ERROR_BUFFER_TOO_SMALL);
         }
         else
         {
-            bbuf.PutLE32((uint32_t) mLength);
-            VerifyOrExit(bbuf.Fit(), err = CHIP_ERROR_BUFFER_TOO_SMALL);
+            bbuf.PutLE32(static_cast<uint32_t>(mLength));
         }
     }
 
-    bbuf.Put(mMetadata, (size_t) mMetadataLength);
-    VerifyOrExit(bbuf.Fit(), err = CHIP_ERROR_BUFFER_TOO_SMALL);
+    bbuf.Put(mMetadata, static_cast<size_t>(mMetadataLength));
 
-    aBuffer->SetDataLength(static_cast<uint16_t>(bbuf.Written()));
+    VerifyOrExit(bbuf.Fit(), err = CHIP_ERROR_BUFFER_TOO_SMALL);
+    aBuffer.SetDataLength(static_cast<uint16_t>(bbuf.Written()));
 
 exit:
     return err;
 }
 
-CHIP_ERROR ReceiveAccept::Parse(System::PacketBuffer * aBuffer, ReceiveAccept & aParsedMessage)
+CHIP_ERROR ReceiveAccept::Parse(System::PacketBuffer & aBuffer, ReceiveAccept & aParsedMessage)
 {
     CHIP_ERROR err          = CHIP_NO_ERROR;
     bool hasStartOffset     = false;
     uint8_t tcByte          = 0;
     uint8_t rangeCtl        = 0;
     uint8_t tmpControlMode  = 0;
-    uint8_t * bufStart      = NULL;
     uint32_t tmpUint32Value = 0;
+    uint8_t * bufStart      = aBuffer.Start();
+    Reader bufReader(bufStart, aBuffer.DataLength());
 
-    // BufBound has to be initialized with aBuffer so VerifyOrExit can't be used here.
-    if (aBuffer == NULL)
-    {
-        return CHIP_ERROR_INVALID_ARGUMENT;
-    }
-
-    bufStart = aBuffer->Start();
-    Reader bufReader(bufStart, aBuffer->DataLength());
-
-    SuccessOrExit(bufReader.Read8(&tcByte).StatusCode());
+    SuccessOrExit(bufReader.Read8(&tcByte).Read8(&rangeCtl).Read16(&aParsedMessage.mMaxBlockSize).StatusCode());
 
     aParsedMessage.mVersion = tcByte & VERSION_MASK;
     tmpControlMode          = tcByte & CONTROL_MODE_MASK;
 
     // Determine transfer control mode. Only one mode should be selected.
-    // TODO: should this verification happen here
+    // TODO: should this verification happen here?
```

Comment:
Open question: should Parse() verify that a single control mode has been specified, or should the caller? 

The argument for not doing it here is that the caller of Parse() will need to interpret the CHIP_ERROR returned in this case and translate it to a BDX profile error that can be sent in a StatusReport.

However, if the caller does the verification, it means that the struct would have to store the transfer control byte raw instead of interpreting it.

Reply:
I think you could extract the SenderDrive/ReceiverDrive/Async bits as bool, and let the caller do the check.

I would also say "CHIP_ERROR_MESSAGE_INCOMPLETE" is likely not the right error if the check is done here.

Reply:
Ok that works. I think that's more in line with the rest of the struct having minimal logic. It should just pack and unpack and not verify the contents beyond whether they're there or not